File: src/main/groovy/shadow/resource/IncludeResourceTransformer.java
package shadow.resource;

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import org.apache.maven.plugins.shade.relocation.Relocator;
import org.codehaus.plexus.util.IOUtil;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.List;
import java.util.jar.JarEntry;
import java.util.jar.JarOutputStream;

/**
 * A resource processor that allows the addition of an arbitrary file
 * content into the shaded JAR.
 */
public class IncludeResourceTransformer
    implements ResourceTransformer
{
    File file;

    String resource;

    public boolean canTransformResource( String r )
    {
        return false;
    }

    public void processResource( String resource, InputStream is, List<Relocator> relocators )
        throws IOException
    {
        // no op
    }

    public boolean hasTransformedResource()
    {
        return file != null ? file.exists() : false;
    }

    public void modifyOutputStream( JarOutputStream jos )
        throws IOException
    {
        jos.putNextEntry( new JarEntry( resource ) );

        InputStream in = new FileInputStream( file );
        IOUtil.copy( in, jos );
        in.close();
    }
}

File: api/src/main/kotlin/com/github/jengelman/gradle/plugins/shadow/transformers/IncludeResourceTransformer.kt
package com.github.jengelman.gradle.plugins.shadow.transformers

import com.github.jengelman.gradle.plugins.shadow.transformers.TransformerContext.Companion.getEntryTimestamp
import java.io.File
import org.apache.tools.zip.ZipEntry
import org.apache.tools.zip.ZipOutputStream
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.InputFile
import org.gradle.api.tasks.PathSensitive
import org.gradle.api.tasks.PathSensitivity

/**
 * A resource processor that allows the addition of an arbitrary file
 * content into the shaded JAR.
 *
 * Modified from `org.apache.maven.plugins.shade.resource.IncludeResourceTransformer.java`
 *
 * @author John Engelman
 */
open class IncludeResourceTransformer : Transformer by NoOpTransformer {
    @get:InputFile
    @get:PathSensitive(PathSensitivity.NONE)
    var file: File? = null

    @get:Input
    var resource: String? = null

    override fun hasTransformedResource(): Boolean = file?.exists() == true

    override fun modifyOutputStream(os: ZipOutputStream, preserveFileTimestamps: Boolean) {
        val entry = ZipEntry(requireNotNull(resource))
        entry.time = getEntryTimestamp(preserveFileTimestamps, entry.time)
        os.putNextEntry(entry)

        requireNotNull(file).inputStream().use { inputStream ->
            inputStream.copyTo(os)
        }
    }
}
